core/src/main/java/hudson/model/Queue.java

```diff
@@ -3009,4 +3013,66 @@ public static Queue getInstance() {
     public static void init(Jenkins h) {
         h.getQueue().load();
     }
+
+    /**
+     * Schedule <tt>Queue.save()</tt> call for near future once items change. Ignore all changes until the time the save
+     * takes place.
+     *
+     * Once queue is restored after a crash, items stages might not be accurate until the next #maintain() - this is not
+     * a problem as the items will be reshuffled first and then scheduled during the next maintainance cycle.
+     *
+     * Implementation note: Queue.load() calls QueueListener hooks for every item deserialized that can hammer the persistance
+     * on load. The problem is avoided by delaying the actual save for the time long enough for queue to load so the save
+     * operations will collapse into one. Also, items are persisted as buildable or blocked in vast majority of cases and
+     * those stages does not trigger the save here.
+     */
+    @Extension
+    @Restricted(NoExternalUse.class)
+    public static final class Saver extends QueueListener implements Runnable {
+
+        /**
+         * All negative values will disable periodic saving.
+         */
+        @VisibleForTesting
+        /*package*/ static /*final*/ int DELAY_SECONDS = SystemProperties.getInteger("hudson.model.Queue.Saver.DELAY_SECONDS", 60);
+
+        private volatile Future<?> nextSave;
```

Comment:
Testcase looks good and simpler, but... you know, I really think you want to be using synchronization rather than volatile here, since a future has its own internal state and we want to ensure only one thread at a time is working with it.   

In particular I'm 95% sure that the `push()` method is actually NOT thread-safe without synchronization since there are separate read and write operations to `nextSave` -- the fact that threads aren't getting a CPU-cache-local versions of the variable (b/c of the `volatile` keyword) won't save us there.

Reply:
Probably for this we want simple synchronization - a `ReadWriteLock` or `StampedLock` would permit uncontended reads but I doubt the extra overhead is worth it when the lock is held so briefly.

Reply:
I do not think `Future` reference itself needs to be used in a synchronized fashion - after all, it is a synchronization primitive. You are right though that `push()` is buggy because of multiple reads of volatile reference.

Reply:
@olivergondza Just synchronize the access to the field, it's too easy to hang yourself with subtle bugs around `volatile` -- the current implementation still doesn't appear to be correct to me.